### ISSUE TITLE
Unexpected higher-ranked lifetime error in GAT usage

### ISSUE THREAD

**Issue by ethe** [rust-lang/rust#100013]
I tried this code:

```rust
#![feature(generic_associated_types)]
#![feature(type_alias_impl_trait)]

use std::future::Future;

pub trait FutureIterator: 'static {
    type Iterator;

    type Future<'s, 'cx>: Future<Output = Self::Iterator> + Send + 'cx
    where
        's: 'cx;

    fn get_iter<'s, 'cx>(&'s self, info: &'cx ()) -> Self::Future<'s, 'cx>;
}

trait IterCaller: 'static {
    type Future1<'cx>: Future<Output = ()> + Send + 'cx;
    type Future2<'cx>: Future<Output = ()> + Send + 'cx;

    fn call_1<'s, 'cx>(&'s self, cx: &'cx ()) -> Self::Future1<'cx>
    where
        's: 'cx;
    fn call_2<'s, 'cx>(&'s self, cx: &'cx ()) -> Self::Future2<'cx>
    where
        's: 'cx;
}

struct UseIter<FI1, FI2> {
    fi_1: FI1,
    fi_2: FI2,
}

impl<FI1, FI2> IterCaller for UseIter<FI1, FI2>
where
    FI1: FutureIterator + 'static + Send + Sync,
    for<'s, 'cx> FI1::Future<'s, 'cx>: Send,
    FI2: FutureIterator + 'static + Send + Sync,
{
    type Future1<'cx> = impl Future<Output = ()> + Send + 'cx
    where
        Self: 'cx;

    type Future2<'cx> = impl Future<Output = ()> + Send + 'cx
    where
        Self: 'cx;

    fn call_1<'s, 'cx>(&'s self, cx: &'cx ()) -> Self::Future1<'cx>
    where
        's: 'cx,
    {
        async {
            self.fi_1.get_iter(cx).await;
        }
    }

    fn call_2<'s, 'cx>(&'s self, cx: &'cx ()) -> Self::Future2<'cx>
    where
        's: 'cx,
    {
        async {
            self.fi_2.get_iter(cx).await;
        }
    }
}

fn main() {}

```

I expected to see this happen: both IterCaller::call_1  and IterCaller::call_2 pass the compiling

Instead, this happened:
```
error: higher-ranked lifetime error
  --> src/main.rs:60:9
   |
60 | /         async {
61 | |             self.fi_2.get_iter(cx).await;
62 | |         }
   | |_________^
```

### Meta
`rustc --version --verbose`:
```
rustc 1.64.0-nightly (f9cba6374 2022-07-31)
binary: rustc
commit-hash: f9cba63746d0fff816250b2ba7b706b5d4dcf000
commit-date: 2022-07-31
host: x86_64-unknown-linux-gnu
release: 1.64.0-nightly
LLVM version: 14.0.6
```

**Comment by aliemjay** [comment]
This is  another manifestation of #92096.

Minimized:
```rust
#![feature(generic_associated_types)]

pub trait FutureIterator {
    type Future<'s, 'cx>: Send
    where
        's: 'cx;
}

fn call_2<I: FutureIterator>() -> impl Send {
    async { // a generator checked for autotrait impl `Send`
        let x = None::<I::Future<'_, '_>>; // a type referencing GAT
        async {}.await; // a yield point
    }
}
```

@rustbot F-generic_associated_types A-async-await A-lifetimes A-generators T-types

**Comment by tmandry** [comment]
This is an unfortunate error message. Since GATs is close to stabilization, does it make sense to nominate for types team discussion?

@rustbot label +I-types-nominated

**Comment by eholk** [comment]
Yeah, I'd like to hear from the types team on this. @nikomatsakis or @jackh726, is adding I-types-nominated the way to do that?

@rustbot label +AsyncAwait-Triaged

**Comment by jackh726** [comment]
This error message sucks, but is likely related to TAITs. Nevertheless, I'm going to take some time this upcoming week (I think) to take a look at least to improve the error message.

**Comment by PureWhiteWu** [comment]
Hello, is there any progress about this?

**Comment by chrysn** [comment]
I'm seeing a similar message when using elaborate GAT-heavy traits; it's far from a minimal example as it is now, but there's no `async` block or function involved. Is there value in attempting to minimize the example, or is the underlying problem of this error appearing without any apparent cause understood well enough already (so that I'd just wait for the fix, and only revisit if this issue gets closed and mine persists)?

**Comment by samlich** [comment]
I'm getting this issue without async, though it gives a `lifetime bound not satisfied`, instead of a higher-ranked lifetime error.

[Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7248905d3cb4abc1884c8b401e146119)

```
trait Tr {
        type A<'a, 'b: 'a>;
}

struct S<T> {
        a: T,
}

// bad
impl<T: Tr> S<T>
where
        for<'a, 'b> T: Tr<A<'a, 'b> = ()>,
        for<'c, 'd> T::A<'c, 'd>: Default,
{
}
```

```
error: lifetime bound not satisfied
  --> src/lib.rs:10:1
   |
10 | / impl<T: Tr> S<T>
11 | | where
12 | |     for<'a, 'b> T: Tr<A<'a, 'b> = ()>,
13 | |     for<'c, 'd> T::A<'c, 'd>: Default,
14 | | {
15 | | }
   | |_^
   |
note: the lifetime `'c` defined here...
  --> src/lib.rs:13:6
   |
13 |     for<'c, 'd> T::A<'c, 'd>: Default,
   |         ^^
note: ...must outlive the lifetime `'d` defined here
  --> src/lib.rs:13:10
   |
13 |     for<'c, 'd> T::A<'c, 'd>: Default,
   |             ^^
   = note: this is a known limitation that will be removed in the future (see issue #100013 <https://github.com/rust-lang/rust/issues/100013> for more information)
```

With nightly 2022-11-08 (before error message change):

```
error[E0478]: lifetime bound not satisfied
  --> src/lib.rs:10:1
   |
10 | / impl<T: Tr> S<T>
11 | | where
12 | |     for<'a, 'b> T: Tr<A<'a, 'b> = ()>,
13 | |     for<'c, 'd> T::A<'c, 'd>: Default,
14 | | {
15 | | }
   | |_^
```

Only happens when both constraints are put on the impl. Also solved by removing the `'b: 'a` constraint on `C`.

Adding `'a: 'b` to `C` causes it to also complain that `'d` must outlive `'c`. I don't understand why they're reversed.

**Comment by tlowerison** [comment]
I've also ran into this issue without async. Interestingly, the correct implied lifetime bound seems to be getting used when a struct is used to impose the bound as opposed to a GAT.

[Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021)
```
struct B<'a: 'b, 'b>(&'b &'a ());

trait Foo {
    type Bar<'a, 'b>
    where
        'a: 'b;

    fn suceeds1<F, T>(&self, f: F) -> T
    where
        F: for<'a, 'b> FnOnce(&'b B<'a, 'b>) -> T;

    fn succeeds2(&self)
    where
        for<'a, 'b> B<'a, 'b>: Default;

    fn fails1<F, T>(&self, f: F) -> T
    where
        F: for<'a, 'b> FnOnce(&'b Self::Bar<'a, 'b>) -> T;

    fn fails2(&self)
    where
        for<'a, 'b> Self::Bar<'a, 'b>: Default;
}

impl Foo for () {
    type Bar<'a, 'b> = B<'a, 'b> where 'a: 'b;

    fn suceeds1<F, T>(&self, f: F) -> T
    where
        F: for<'a, 'b> FnOnce(&'b B<'a, 'b>) -> T,
    {
        todo!()
    }

    fn succeeds2(&self)
    where
        for<'a, 'b> B<'a, 'b>: Default,
    {
        todo!()
    }

    fn fails1<F, T>(&self, f: F) -> T
    where
        F: for<'a, 'b> FnOnce(&'b Self::Bar<'a, 'b>) -> T,
    {
        todo!()
    }

    fn fails2(&self)
    where
        for<'a, 'b> Self::Bar<'a, 'b>: Default,
    {
        todo!()
    }
}
```
With `nightly-2023-03-07-aarch64-apple-darwin` toolchain and on playground I get
```
error: lifetime bound not satisfied
  --> src/lib.rs:42:8
   |
42 |     fn fails1<F, T>(&self, f: F) -> T
   |        ^^^^^^
   |
note: the lifetime `'b` defined here...
  --> src/lib.rs:44:20
   |
44 |         F: for<'a, 'b> FnOnce(&'b Self::Bar<'a, 'b>) -> T,
   |                    ^^
note: ...must outlive the lifetime `'a` defined here
  --> src/lib.rs:44:16
   |
44 |         F: for<'a, 'b> FnOnce(&'b Self::Bar<'a, 'b>) -> T,
   |                ^^
   = note: this is a known limitation that will be removed in the future (see issue #100013 <https://github.com/rust-lang/rust/issues/100013> for more information)

error: lifetime bound not satisfied
  --> src/lib.rs:49:8
   |
49 |     fn fails2(&self)
   |        ^^^^^^
   |
note: the lifetime `'b` defined here...
  --> src/lib.rs:51:17
   |
51 |         for<'a, 'b> Self::Bar<'a, 'b>: Default,
   |                 ^^
note: ...must outlive the lifetime `'a` defined here
  --> src/lib.rs:51:13
   |
51 |         for<'a, 'b> Self::Bar<'a, 'b>: Default,
   |             ^^
   = note: this is a known limitation that will be removed in the future (see issue #100013 <https://github.com/rust-lang/rust/issues/100013> for more information)

error: could not compile `gat-lifetime-error` due to 2 previous errors
```

**Comment by urben1680** [comment]
This code is also triggering the compilation error that links to this issue:

```rs
use core::marker::PhantomData;
struct Omg<'a>(PhantomData<&'a mut ()>);

trait Foo {
    type Bar<'a: 'b, 'b>;
    fn baz<'a>(f: impl for<'b> Fn(Self::Bar<'a, 'b>));
}

impl Foo for f32 {
    type Bar<'a: 'b, 'b> = &'b mut Omg<'a>;
    fn baz //error[E0478]: lifetime bound not satisfied
    (_f: impl for <'a, 'b> Fn(Self::Bar<'a, 'b>)){}
}
```

**Comment by nandesu-utils** [comment]
A weird RPITIT error with lifetime bounds. This one fails to borrow check with outlives-bound.
```rust
#![feature(return_position_impl_trait_in_trait)]

use std::future::Future;

pub trait Foo: Sync {
    fn run<'a, 'b: 'a, T: Sync>(&'a self, _: &'b T) -> impl Future<Output = ()> + 'a + Send; // doesn't compile
    // fn run<'a, 'b, T: Sync>(&'a self, _: &'b T) -> impl Future<Output = ()> + 'a + Send; // compiles
}

pub trait FooExt: Foo {
    fn run_via<'a, 'b: 'a, T: Sync>(&'a self, t: &'b T) -> impl Future<Output = ()> + 'a + Send {
        async move {
            // asks for an unspecified lifetime to outlive itself? weird diagnostics
            self.run(t).await;
        }
    }
}
```

However I think this is really weird that if we remove `Send` bounds everywhere in RPITITs, it will work.
```rust
pub trait Foo: Sync {
    fn run<'a, 'b: 'a, T: Sync>(&'a self, _: &'b T) -> impl Future<Output = ()> + 'a;
}

pub trait FooExt: Foo {
    fn run_via<'a, 'b: 'a, T: Sync>(&'a self, t: &'b T) -> impl Future<Output = ()> + 'a {
        async move {
            self.run(t).await; // compiles now!
        }
    }
}
```

Seems like proving `Send` breaks it. A type-erased `Pin<Box<...>>` (no RPITITs) works flawlessly however.

**Comment by compiler-errors** [comment]
@nandesu-utils: I don't think this error is related. Do you mind filing a separate bug for that?

**Comment by nandesu-utils** [comment]
> @nandesu-utils: I don't know if this error is related. Do you mind filing a separate bug for that?

Surely! I hope I didn't mess it up. #111105

**Comment by Dessix** [comment]
Having read over the description @nandesu-utils gave around "Send", this may also share a root-cause with #102211, as they both appear to rest in propagation of `Send` provability.

**Comment by PureWhiteWu** [comment]
Hi, is there any progress on this?
Since #115822 is stabilized, we meets this problem that prevents us from migrating to RPITIT.

**Comment by PureWhiteWu** [comment]
A new reproduce of this, but not quite sure if this is the same problem:
```rust
use std::future::Future;

pub trait Service {
    fn call<'s: 'cx, 'cx>(&self) -> impl Future<Output = ()> + Send;
}

struct ServiceImpl<S> {
    inner: S,
}

impl<S> Service for ServiceImpl<S>
where
    S: Service + Sync + Send,
{
    fn call<'s: 'cx, 'cx>(&'s self) -> impl Future<Output = ()> {
        async move {
            self.inner.call().await;
        }
    }
}
```

Playground link: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=44ba55e3e70014c1eab369f0191bb647

**Comment by kellpossible** [comment]
I've been unable to use https://github.com/a1akris/page-turner/ version 1.0.0 without encountering this error.

My current workaround is wrapping the pages stream in:

```rust
pub fn assert_send_stream<R>(
    s: impl futures::Stream<Item = R> + Send,
) -> impl futures::Stream<Item = R> + Send {
    s
}
```

**Comment by AlirezaDarbandi** [comment]
> I tried this code:
> 
> ```rust
> #![feature(generic_associated_types)]
> #![feature(type_alias_impl_trait)]
> 
> use std::future::Future;
> 
> pub trait FutureIterator: 'static {
>     type Iterator;
> 
>     type Future<'s, 'cx>: Future<Output = Self::Iterator> + Send + 'cx
>     where
>         's: 'cx;
> 
>     fn get_iter<'s, 'cx>(&'s self, info: &'cx ()) -> Self::Future<'s, 'cx>;
> }
> 
> trait IterCaller: 'static {
>     type Future1<'cx>: Future<Output = ()> + Send + 'cx;
>     type Future2<'cx>: Future<Output = ()> + Send + 'cx;
> 
>     fn call_1<'s, 'cx>(&'s self, cx: &'cx ()) -> Self::Future1<'cx>
>     where
>         's: 'cx;
>     fn call_2<'s, 'cx>(&'s self, cx: &'cx ()) -> Self::Future2<'cx>
>     where
>         's: 'cx;
> }
> 
> struct UseIter<FI1, FI2> {
>     fi_1: FI1,
>     fi_2: FI2,
> }
> 
> impl<FI1, FI2> IterCaller for UseIter<FI1, FI2>
> where
>     FI1: FutureIterator + 'static + Send + Sync,
>     for<'s, 'cx> FI1::Future<'s, 'cx>: Send,
>     FI2: FutureIterator + 'static + Send + Sync,
> {
>     type Future1<'cx> = impl Future<Output = ()> + Send + 'cx
>     where
>         Self: 'cx;
> 
>     type Future2<'cx> = impl Future<Output = ()> + Send + 'cx
>     where
>         Self: 'cx;
> 
>     fn call_1<'s, 'cx>(&'s self, cx: &'cx ()) -> Self::Future1<'cx>
>     where
>         's: 'cx,
>     {
>         async {
>             self.fi_1.get_iter(cx).await;
>         }
>     }
> 
>     fn call_2<'s, 'cx>(&'s self, cx: &'cx ()) -> Self::Future2<'cx>
>     where
>         's: 'cx,
>     {
>         async {
>             self.fi_2.get_iter(cx).await;
>         }
>     }
> }
> 
> fn main() {}
> 
> ```
> 
> I expected to see this happen: both type declaration of `IterCaller::call_1`  and `IterCaller::call_2` pass the compiling
> 
> Instead, this happened:
> ```
> error: higher-ranked lifetime error
>   --> src/main.rs:60:9
>    |
> 60 | /         async {
> 61 | |             self.fi_2.get_iter(cx).await;
> 62 | |         }
>    | |_________^
> ```
> 
> ### Meta
> `rustc --version --verbose`:
> ```
> rustc 1.64.0-nightly (f9cba6374 2022-07-31)
> binary: rustc
> commit-hash: f9cba63746d0fff816250b2ba7b706b5d4dcf000
> commit-date: 2022-07-31
> host: x86_64-unknown-linux-gnu
> release: 1.64.0-nightly
> LLVM version: 14.0.6
> ```
>

**Comment by nitsky** [comment]
I recently encountered this issue with the following code:

```rust
use std::future::Future;

trait Trait: Sync {
    type T<'a>: Default;
    fn f(&self, t: Self::T<'_>) -> impl Future<Output = ()> + Send;
}

fn g(t: &impl Trait) -> impl Future<Output = ()> + Send + '_ {
    async {
        t.f(Default::default()).await
    }
}
```

The error can be avoided with any of the following changes:

1. Removing the lifetime from the associated type.
2. Removing the Send bound from the trait function return type.
3. Returning the future directly rather than awaiting it in an async block.

In my application, I found it easiest to work around the issue by boxing the future.

```diff
-        t.f(Default::default()).await
+        use futures::FutureExt;
+        t.f(Default::default()).boxed().await
```

Note that the `'_` bound will no longer be necessary with the 2024 capture rules.

**Comment by rvolosatovs** [comment]
> I recently encountered this issue with the following code:
> 
> ```rust
> use std::future::Future;
> 
> trait Trait: Sync {
>     type T<'a>: Default;
>     fn f(&self, t: Self::T<'_>) -> impl Future<Output = ()> + Send;
> }
> 
> fn g(t: &impl Trait) -> impl Future<Output = ()> + Send + '_ {
>     async {
>         t.f(Default::default()).await
>     }
> }
> ```
> 
> The error can be avoided with any of the following changes:
> 
>     1. Removing the lifetime from the associated type.
> 
>     2. Removing the Send bound from the return type.
> 
>     3. Returning the future directly rather than awaiting it in an async block.
> 
> 
> In my application, I found it easiest to work around the issue by boxing the future.
> 
> ```diff
> -        t.f(Default::default()).await
> +        use futures::FutureExt;
> +        t.f(Default::default()).boxed().await
> ```
> 
> Note that the `'_` bound will no longer be necessary with the 2024 capture rules.

thanks @nitsky, your `boxed()` observation really helped me to debug this and I found a workaround that does not require a heap allocation:

```rust
pub trait SendFuture: core::future::Future {
    fn send(self) -> impl core::future::Future<Output = Self::Output> + Send
    where
        Self: Sized + Send,
    {
        self
    }
}

impl<T: core::future::Future> SendFuture for T {}
```

my reproducer looks like the following:

```rust
    trait X {
        fn test<Y>(&self, x: impl AsRef<[Y]>) -> impl core::future::Future<Output = ()> + Send
        where
            Y: AsRef<str>;
    }

    fn call_test(x: impl X + Send + Sync) -> impl core::future::Future<Output = ()> + Send {
        async move { x.test(["test"]).await }
        // fix:
        // use send_future::SendFuture as _;
        // async move { x.test(["test"]).send().await }
    }
```

I've published this trait in `send-future` crate with docs available here: https://docs.rs/send-future/latest/send_future/trait.SendFuture.html

**Comment by kanpov** [comment]
This error pops up for me in the following context:

```rust
pub async fn read_pid(filesystem: &impl LinuxFilesystem, pid_file: String) -> Option<u32> {
    let pid_path = OsString::from(pid_file);
    let pid_path = pid_path.as_os_str();
    #[allow(unused)]
    let mut pid_option = None;

    loop {
        if let Ok(reader) = filesystem.open_file(pid_path, &LinuxOpenOptions::new().read()).await {
            tokio::pin!(reader);

            let mut content = String::new();
            if reader.read_to_string(&mut content).await.is_err() {
                continue;
            }

            pid_option = content.parse().ok();
            if pid_option.is_some() {
                break;
            }
        }
    }

    pid_option
}
```

And when calling that async fn with a correct self as the first argument. Is there any way to work around this?

**Comment by kanpov** [comment]
> > This error pops up for me in the following context:
> > ```rust
> > pub async fn read_pid(filesystem: &impl LinuxFilesystem, pid_file: String) -> Option<u32> {
> >     let pid_path = OsString::from(pid_file);
> >     let pid_path = pid_path.as_os_str();
> >     #[allow(unused)]
> >     let mut pid_option = None;
> > 
> >     loop {
> >         if let Ok(reader) = filesystem.open_file(pid_path, &LinuxOpenOptions::new().read()).await {
> >             tokio::pin!(reader);
> > 
> >             let mut content = String::new();
> >             if reader.read_to_string(&mut content).await.is_err() {
> >                 continue;
> >             }
> > 
> >             pid_option = content.parse().ok();
> >             if pid_option.is_some() {
> >                 break;
> >             }
> >         }
> >     }
> > 
> >     pid_option
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > And when calling that async fn with a correct self as the first argument. Is there any way to work around this? Without send-future, I get 2 errors: this one and "implementation of Send is not general enough"
> 
> It looks like a quick fix could be wrapping the `filesystem` parameter type into an `Arc` to get rid of the lifetime?
> 
> I also found that implementing the future manually returning an `impl Future` usually helps (and if you do you might want to use the `Captures` trait bound trick for the borrows)

The arc approach is not possible due to the method calling this function accepting &self on a non-arced instance per design.

Implementing the future manually was nightmarish and caused a barrage of issues: async block cannot be Send due to `content` being held across an await point plus lifetime issues, unfortunately it's not possible to not use the `&impl LinuxFilesystem` reference in my case.

**Comment by rvolosatovs** [comment]
@kanpov your original example seems to work in playground https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e816de3c132cfd2563821a4cc9e90ef3

if you use any generic types in return or parameter position, you probably want to mark them `Send` and `'static`

**Comment by kanpov** [comment]
> @kanpov your original example seems to work in playground https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e816de3c132cfd2563821a4cc9e90ef3
> 
> if you use any generic types in return or parameter position, you probably want to mark them `Send` and `'static`

Yes, it compiles but actually calling the function in a `&self` function while passing in `self` produces the "lifetime bound not satisfied" and "implementation of Send is not general enough" errors.

**Comment by ethe** [comment]
Hi all, since this issue is referenced in the rustc error message, I think the title "Unexpected higher-ranked lifetime error in GAT unsafe" is not clear enough. Shall we edit the title of this issue to make it clearer?

**Comment by bjackman** [comment]
I'm running into this error message and minimized my code to [this example](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=483f730e1adaf1fb5302da8152e11110):

```rust
pub struct MyStruct {}

pub trait MyTrait {
    async fn inner<'a, I, >(&self, _args: I)
    where
        I: IntoIterator<Item = &'a str>,
    {
    }

    async fn outer(&self) {
        self.inner(["gud"]).await;
    }
}

impl MyTrait for MyStruct {}

#[tokio::main]
async fn main() {
    let s = MyStruct {};

    tokio::spawn(async move {
        s.outer().await;
    });
}
```

In my case, I did not get the link to this issue (only the "one type is more general than the other") with [the original version](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=ff7568d1e68e5093b2376adafce30cfe), which used `AsRef<str>` instead of `&str`.

This also seems to be a genuinely minimal repro, I cannot reproduce it without both `inner` and `outer` methods.

**Comment by Daniel-Aaron-Bloom** [comment]
I don't know if it helps, but you can repro it [even more minimally](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=caab9b82085bc0863a26e6fc19488766)

```rust
pub trait MyTrait: Sync {
    fn inner<'a>(&self) -> impl Future<Output = ()> + Send;
}

fn outer<T: MyTrait>(t: &T) -> impl Future<Output = ()> + Send {
    async { t.inner().await; }
}
```

Still needs `inner` and `outer`, but here removing the unused `<'a>` erases the error, which is pretty strange.

**Comment by zeenix** [comment]
FWIW, this issue forced [a very major design change](https://github.com/zeenix/zlink/commit/2491fea18c6a853df93ca287ba3d8da9a01e18c3) in a project I'm developing. 😢

**Comment by compiler-errors** [comment]
As a note, I encourage anyone to test whether `-Zhigher-ranked-assumptions` fixes their bugs here:

```
env RUSTFLAGS=-Zhigher-ranked-assumptions cargo +nightly build
```

**Comment by ethe** [comment]
> As a note, I encourage anyone to test whether `-Zhigher-ranked-assumptions` fixes their bugs here:
> 
> ```
> env RUSTFLAGS=-Zhigher-ranked-assumptions cargo +nightly build
> ```

refer to #143545

**Comment by jakob-lilliemarck** [comment]
> > As a note, I encourage anyone to test whether `-Zhigher-ranked-assumptions` fixes their bugs here:
> > ```
> > env RUSTFLAGS=-Zhigher-ranked-assumptions cargo +nightly build
> > ```
> 
> refer to [#143545](https://github.com/rust-lang/rust/pull/143545)

Sure did it for me!

**Comment by jakob-lilliemarck** [comment]
> As a note, I encourage anyone to test whether `-Zhigher-ranked-assumptions` fixes their bugs here:
> 
> ```
> env RUSTFLAGS=-Zhigher-ranked-assumptions cargo +nightly build
> ```

Thats amazing, it resolved the error message for me - any clue how confident I can feel about this flag..?

**Comment by theemathas** [comment]
Another way to trigger this error:

```rust
trait Trait<'b>: 'b {}

fn wut()
where
    for<'a, 'b> dyn Trait<'b> + 'a:,
{
}

fn main() {
    wut();
}
```

```
error: lifetime bound not satisfied
  --> src/main.rs:10:5
   |
10 |     wut();
   |     ^^^^^
   |
note: the lifetime `'b` defined here...
  --> src/main.rs:5:13
   |
 5 |     for<'a, 'b> dyn Trait<'b> + 'a:,
   |             ^^
note: ...must outlive the lifetime `'a` defined here
  --> src/main.rs:5:9
   |
 5 |     for<'a, 'b> dyn Trait<'b> + 'a:,
   |         ^^
   = note: this is a known limitation that will be removed in the future (see issue #100013 <https://github.com/rust-lang/rust/issues/100013> for more information)
```

I'm not sure if this should compile or not.

See also https://github.com/rust-lang/rust/issues/104684#issuecomment-3875210872
